### PR TITLE
[core] Fix GCS client retry

### PR DESCRIPTION
**Pull request by vitsai**
Recreate the grpc channel during retries, in case the originally created channel was faulty.

## Why are these changes needed?

ray start was seeing a 3x slowdown due to failing reconnects on a faulty channel.

## Related issue number

https://github.com/ray-project/ray/issues/38509

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by fishbone — COMMENTED** [review]
I think it should be fine, but @rkooo567 @jjyao  could you give it another look?

**Comment by jjyao** [comment]
What's the before and after time for `ray start --head`?

File: python/ray/_private/node.py
Comment (jjyao) [inline python/ray/_private/node.py:1239]: I think it's fragile: there is a hidden assumption here that monitor and ray client server don't need to access cluster info populated by `_write_cluster_info_to_kv`

Comment (vitsai) [reply python/ray/_private/node.py:1240]: So I guess a more enduring fix might look something like: let GCS serve the kv store before the rest of the server has fully initialized, but there's the whole problem of GCS being single-threaded

File: python/ray/_private/node.py
Comment (vitsai) [reply python/ray/_private/node.py:1055]: `get_gcs_client` will call `_init_gcs_client` the first time it's called

File: python/ray/_private/node.py
Comment (vitsai) [reply python/ray/_private/node.py:1240]: passing the metadata to the processes when they're started is not the best solution, but it's /a/ solution

**Comment by vitsai** [comment]
Time has gone back down from 5.09s to 1.05s

**Comment by vitsai** [comment]
Moved the sync point back to where it was and fixed a bug in the retry logic. It's now 1.97 seconds, which is around where it was originally. Note that if we move the sync point to later in time as the earlier revision did, we can further optimize to ~1.05 at the expense of brittleness.

File: python/ray/_private/node.py
Comment (jjyao) [inline python/ray/_private/node.py:1135]: These are no longer needed

**Review by jjyao — COMMENTED** [review]
I like the new fix :)

File: src/ray/gcs/gcs_client/gcs_client.cc
Comment (jjyao) [inline src/ray/gcs/gcs_client/gcs_client.cc:191]: What's the root cause?

Comment (vitsai) [reply src/ray/gcs/gcs_client/gcs_client.cc:191]: The channel was not being recreated during retries, so an RPC that failed the first time was doomed to fail for every subsequent retry (spaced 1s apart) since a faulty channel would be returned. We also don't check whether the channel is faulty when we first create it.

Comment (vitsai) [reply src/ray/gcs/gcs_client/gcs_client.cc:191]: removed

File: python/ray/_private/node.py
Comment (vitsai) [reply python/ray/_private/node.py:1135]: removed

File: python/ray/scripts/scripts.py
Comment (rkooo567) [inline python/ray/scripts/scripts.py:942]: We should remove this before we merge it. Or we can add the second information around the starting log that we already have

File: src/ray/gcs/gcs_client/gcs_client.cc
Comment (rkooo567) [reply src/ray/gcs/gcs_client/gcs_client.cc:191]: Can we just use https://github.com/ray-project/ray/pull/39017 (cc @rynewang)?

File: src/ray/gcs/gcs_client/gcs_client.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_client/gcs_client.cc:200]: why do we not reset things like `node_info_stub_`?

**Comment by rkooo567** [comment]
Please rewrite the PR title & description!

File: src/ray/gcs/gcs_client/gcs_client.cc
Comment (vitsai) [reply src/ray/gcs/gcs_client/gcs_client.cc:191]: My understanding is that the linked PR is for connection fluctuations on a working channel, whereas in this case, it's not clear that the channel created is working at all (so it is necessary to create a new one). i.e. RPCs fail on the bad channel for 5 seconds, whereas the server comes up within a second and a half -- so wait_for_ready will either be slower than recreating, or it will never work.

File: src/ray/gcs/gcs_client/gcs_client.cc
Comment (vitsai) [reply src/ray/gcs/gcs_client/gcs_client.cc:200]: We initialize `node_info_stub_` at the time of creating the channel, and reset it within the loop, so by this point it should already be initialized. The remaining stubs here have not been initialized, so we do that here.

File: python/ray/scripts/scripts.py
Comment (vitsai) [reply python/ray/scripts/scripts.py:942]: Moved it to the starting log.

File: src/ray/gcs/gcs_client/gcs_client.cc
Comment (rkooo567) [reply src/ray/gcs/gcs_client/gcs_client.cc:191]: I see, that makes sense. So this is the case retry is better than waiting until channel is recovered from a bad state. 

Bad channel -> is it because the channel may never recover because of GCS restart?

File: python/ray/scripts/scripts.py
Comment (rkooo567) [reply python/ray/scripts/scripts.py:942]: I don't see the change yet?

File: src/ray/gcs/gcs_client/gcs_client.cc
Comment (vitsai) [reply src/ray/gcs/gcs_client/gcs_client.cc:191]: Not sure whether the state on the channel is laggy (by several seconds), or if the created channel is just permanently bad (maybe because GCS was not available at time of creation, although in this case it's still starting up, not restarting)

File: python/ray/scripts/scripts.py
Comment (vitsai) [reply python/ray/scripts/scripts.py]: sorry, pushed after responding, should be there now

File: src/ray/gcs/gcs_client/gcs_client.cc
Comment (rkooo567) [reply src/ray/gcs/gcs_client/gcs_client.cc:191]: So the laggy case could be handled by wait_for_ready (the transient state should be recovered), but the closed channel (without proper shutdown) cannot be handled by this IIUC? Anyway it makes sense to retry here.

Btw, @rynewang don't we still need wait_for_ready here still at least for a few seconds in case it takes time to establish the channel? It seems like this approach won't work if the channel connection cannot be estabilshed within 1 second?

File: src/ray/gcs/gcs_server/gcs_server.h
Comment (rynewang) [inline src/ray/gcs/gcs_server/gcs_server.h:101]: const & ?

File: src/ray/gcs/gcs_client/gcs_client.cc
Comment (rynewang) [reply src/ray/gcs/gcs_client/gcs_client.cc:191]: This confuses me a bit. My understanding is a channel can always be created, albeit in a TRANSIENT_FAILURE state if remote is not accessible, and will retry to connect if in that state according to the [grpc doc](https://grpc.github.io/grpc/cpp/md_doc_connectivity-semantics-and-api.html). What does it mean by `since a faulty channel would be returned` and `doomed to fail for every subsequent retry`? I guess with [a wait for retry PR](https://github.com/ray-project/ray/pull/39017) it should work.

File: src/ray/gcs/gcs_server/gcs_server.h
Comment (vitsai) [reply src/ray/gcs/gcs_server/gcs_server.h:101]: I'll change it, but with RVO it shouldn't matter very much

File: python/ray/_private/node.py
Comment (jjyao) [inline python/ray/_private/node.py:1055]: The above comment needs to be removed as well.

File: python/ray/scripts/scripts.py
Comment (jjyao) [inline python/ray/scripts/scripts.py:775]: Don't change the public message

**Review by jjyao — APPROVED** [review]

File: src/ray/gcs/gcs_client/gcs_client.cc
Comment (rkooo567) [reply src/ray/gcs/gcs_client/gcs_client.cc:191]: This makes sense to me. Btw, what's the reason why the first channel is not ready in the first place? Is it because we try connecting to GCS before it is ready? (and exponential backoff will make it longer time to make transient failure -> ready state?)

**Review by rkooo567 — CHANGES_REQUESTED** [review]
LGTM. Before we merge, Should we add some sort of test to measure ray start latency? To catch issues like this in the future? (I will approve it once it is resolved).

**Comment by vitsai** [comment]
The value is pretty platform dependent; 6s -> 2s was on some AWS machine, but on Mac it is 10.7s -> 6.7s. Seems like a test here would just be flaky and also not add much value since it would have to test the ceiling of all the start times. I guess we could choose one specific platform + machine to always run the test on?

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
@vitsai let's remove it from CI test and add it to microbenchmark. Let's do it in a follow up PR

**Comment by rkooo567** [comment]
premrege failure

**Comment by rkooo567** [comment]
Will merge once it passes https://buildkite.com/ray-project/release-tests-pr/builds/51818

**Comment by rkooo567** [comment]
Passes
```


[INFO 2023-09-01 06:01:41,857] anyscale_job_runner.py: 145  Output: {'collected_metrics': True, 'last_prepare_time_taken': 6.347128261999956, 'prepare_return_codes': [0], 'return_code': 0, 'total_time_taken': 360.87290291500005, 'uploaded_artifact': False, 'uploaded_metrics': True, 'uploaded_results': True, 'workload_time_taken': 307.565308664}
--
  |  

<br class="Apple-interchange-newline">
```

**Comment by rickyyx** [comment]
I think this also fixes https://github.com/ray-project/ray/issues/39111 

I ran the test 100 times on this PR, and it now only failed 4 out of 100 times: https://buildkite.com/ray-project/oss-ci-build-pr/builds/34672#018a4def-7346-4291-8285-34cbaa818016
image

**Pull request merged**
